build/gen_stub.php

```diff
@@ -1513,48 +1481,219 @@ public function getFieldSynopsisElement(DOMDocument $doc): DOMElement
+        $className = str_replace("\\", "-", strtolower($class));
+        $varnameElement = $doc->createElement("varname", $this->getVariableLikeName());
+        $varnameElement->setAttribute("linkend", "$className." . ($isConst ? "constants" : "props") . "." . strtolower($this->getVariableLikeName()));
         $fieldsynopsisElement->appendChild(new DOMText("\n     "));
         $fieldsynopsisElement->appendChild($varnameElement);
 
-        if ($this->defaultValueString) {
+        if ($valueString) {
             $fieldsynopsisElement->appendChild(new DOMText("\n     "));
-            $initializerElement = $doc->createElement("initializer",  $this->defaultValueString);
+            $initializerElement = $doc->createElement("initializer",  $valueString);
             $fieldsynopsisElement->appendChild($initializerElement);
         }
 
         $fieldsynopsisElement->appendChild(new DOMText("\n    "));
 
         return $fieldsynopsisElement;
     }
+}
 
-    /** @return mixed */
-    private function evaluateDefaultValue(bool &$defaultValueConstant)
+class ClassConstInfo extends VariableLike
+{
+    /** @var ClassConstName */
+    public $name;
+    /** @var Expr|null */
+    public $value;
+    /** @var bool */
+    public $isDeprecated;
+    /** @var string|null */
+    public $valueString;
+    /** @var string|null */
+    public $cond;
+
+    public function __construct(ClassConstName $name, int $flags, ?Expr $value, ?string $valueString, bool $isDeprecated, ?string $cond)
     {
-        $evaluator = new ConstExprEvaluator(
-            function (Expr $expr) use (&$defaultValueConstant) {
-                if ($expr instanceof Expr\ConstFetch) {
-                    $defaultValueConstant = true;
-                    return null;
-                }
+        $this->name = $name;
+        $this->flags = $flags;
+        $this->value = $value;
+        $this->valueString = $valueString;
+        $this->isDeprecated = $isDeprecated;
+        $this->cond = $cond;
+    }
 
-                throw new Exception("Property $this->name has an unsupported default value");
+    protected function getVariableTypeCode(): string
+    {
+        return "const";
+    }
+
+    protected function getVariableTypeName(): string
+    {
+        return "class constant";
+    }
+
+    protected function getVariableLikeName(): string
+    {
+        return $this->name->const;
+    }
+
+    public function getDeclaration(): string {
+        $code = "\n";
+
+        $constName = $this->name->const;
+
+        $constantValue = false;
+        if ($this->value === null) {
+            $value = null;
+            $valueType = "undefined";
+        } else {
+            $value = $this->evaluateValue($this->value, $constantValue);
+            $valueType = gettype($value);
+        }
+
+        if ($constantValue) {
+            echo "Skipping code generation for constant $this->name, because it has a constant value\n";
```

Comment:
We could ease this limitation if we built a constant map. Unexposed constant values would be still a problem though, just like the usage of `ZEND_ACC_*` in ext/reflection. E.g. we don't want to generate `ZEND_ACC_PUBLIC` from stubs.